### PR TITLE
mingw: Use aro instead of clang for preprocessing import libs

### LINKED ISSUE #17753: make mingw .def.in file parsing use aro's preprocessor instead of clang
This code converts mingw's .def.in files to .def files:

https://github.com/ziglang/zig/blob/1c85b0acbb5148dec7ff671e767354f3df5c6b83/src/mingw.zig#L291-L293

https://github.com/ziglang/zig/blob/1c85b0acbb5148dec7ff671e767354f3df5c6b83/src/mingw.zig#L348-L375

Instead of using clang's preprocessor, it should use Aro's.

Related:
* #17752

cc @Vexu @ehaas

### PR DESCRIPTION
**Pull request by ehaas**
Based on manual testing, this produces the same output for `kernel32.def` when compared to clang, other than minor whitespace differences (clang puts a newline at the beginning, and a trailing space on one line which aro does not)

Testing steps:
1. Clear zig cache
2. Create empty `test.c`
3. With status-quo zig: `zig build-lib -target x86_64-windows-gnu -lkernel32 --verbose-cc test.c` 
a. This will print a `zig clang` line to stderr; the output path is the final argument
4. With this branch: `./zig-out/bin/zig build-lib -target x86_64-windows-gnu -lkernel32 --verbose-cc test.c`
a. This will print `output path: <ZIG_CACHE_PATH>/o/<SOME HASH>/kernel32.def`
5. Diff the two files 
6. Available targets are `x86_64-windows-gnu`, `x86-windows-gnu`, `arm-windows-gnu`, `aarch64-windows-gnu`

I manually tested error handling by editing `kernel32.def.in` to have an invalid preprocessing directive and got the following:
```
zig-out/lib/zig/libc/mingw/lib-common/kernel32.def.in:1:1: error: invalid preprocessing directive
#foo
^
1 error generated.
error: unable to generate DLL import .lib file for kernel32: AroPreprocessorFailed
```

Closes #17753

### CODE REVIEW

**Comment by andrewrk** [comment]
I suggest to put a check for `build_options.only_core_functionality` to prevent Aro from being compiled into zig1.wasm or zig2.c during the bootstrap process. This will also fix the current CI failures.

**Comment by ehaas** [comment]
Is there a way to test this locally if I don't have access to a windows machine? I don't want to spam the CI with attempts, since I'm not totally sure where the check should go. It seems like the function is being used, since we're hitting that panic - does that mean that if `build_options.only_core_functionality` is true, we should use the existing `zig clang` solution, and otherwise use the new aro one?

**Comment by andrewrk** [comment]
Yes, here is a way to test it:

1. put  `@compileError()` somewhere inside aro
2. `zig build -Donly-c`

If you get the compile error, it's incorrectly including aro in the build.

The check goes at the top of `buildImportLib`:

```zig
if (build_options.only_core_functionality) @panic("building import libs not included in core functionality");
```

**Comment by andrewrk** [comment]
Oops, I gave you some bad advice. It should be checking `only_c` instead of `only_core_functionality`.

We don't want the logic to be part of zig1.wasm, however we do want the logic to be part of zig2.c.

The use case for `only_core_functionality` is for example package fetching. We certainly do not need zig2 to have the package manager logic inside of it.

File: src/mingw.zig [hunk 1]
```diff
@@ -305,6 +302,7 @@ pub fn buildImportLib(comp: *Compilation, lib_name: []const u8) !void {
     cache.hash.addBytes(build_options.version);
     cache.hash.addOptionalBytes(comp.zig_lib_directory.path);
     cache.hash.add(target.cpu.arch);
+    cache.hash.addBytes("aro");
```
Comment (andrewrk) [inline src/mingw.zig:305]: No need, already covered by `build_options.version`.

**Comment by ehaas** [comment]
Ok, if I add that I assume it means it's also safe for me to remove the `if (builtin.zig_backend == .stage2_c) @panic("the CBE cannot compile Aro yet!");` from later in the function, right before I import aro?

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/mingw.zig [hunk 2]
```diff
@@ -337,74 +335,66 @@ pub fn buildImportLib(comp: *Compilation, lib_name: []const u8) !void {
...
+        defer def_final_file.close();
+        try pp.prettyPrintTokens(def_final_file.writer());
     }
 
     const lib_final_path = try comp.global_cache_directory.join(comp.gpa, &[_][]const u8{
         "o", &digest, final_lib_basename,
     });
     errdefer comp.gpa.free(lib_final_path);
 
+    if (!build_options.have_llvm) return error.ZigCompilerNotBuiltWithLLVMExtensions;
```
[the remark this answers is not in the document]
Comment (andrewrk) [reply src/mingw.zig:395]: #17807

File: src/mingw.zig [hunk 3]
```diff
@@ -337,74 +335,66 @@ pub fn buildImportLib(comp: *Compilation, lib_name: []const u8) !void {
...
-        "-undef",
-        "-P",
-        "-I",
-        try comp.zig_lib_directory.join(arena, &[_][]const u8{ "libc", "mingw", "def-include" }),
-        target_def_arg,
-        "-E",
-        "-o",
-        def_final_path,
-    };
+    if (builtin.zig_backend == .stage2_c) @panic("the CBE cannot compile Aro yet!");
```
[the remark this answers is not in the document]
Comment (ehaas) [reply src/mingw.zig]: I get a bunch of errors when compiling zig2.c - the first few:

```
/Users/ehaas/source/zig/build/zig2.c:799577:1154: error: call to undeclared function 'zig_make_big'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
/Users/ehaas/source/zig/build/zig2.c:799577:1183: error: call to undeclared function 'zig_shl_big'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
/Users/ehaas/source/zig/build/zig2.c:799577:1183: note: did you mean 'zig_shl_i8'?
/Users/ehaas/source/zig/stage1/zig.h:534:1: note: 'zig_shl_i8' declared here
zig_int_helpers(8)
^
/Users/ehaas/source/zig/stage1/zig.h:464:5: note: expanded from macro 'zig_int_helpers'
    zig_shift_operator( int##w##_t, shl_i##w, <<) \
    ^
/Users/ehaas/source/zig/stage1/zig.h:455:5: note: expanded from macro 'zig_shift_operator'
    zig_operator(Type, uint8_t, operation, operator)
    ^
/Users/ehaas/source/zig/stage1/zig.h:449:24: note: expanded from macro 'zig_operator'
    static inline Type zig_##operation(Type lhs, RhsType rhs) { \
                       ^
<scratch space>:93:1: note: expanded from here
zig_shl_i8
^
/Users/ehaas/source/zig/build/zig2.c:799577:1236: error: too few arguments to function call, expected 5, have 2
/Users/ehaas/source/zig/stage1/zig.h:1996:20: note: 'zig_or_big' declared here
static inline void zig_or_big(void *res, const void *lhs, const void *rhs, bool is_signed, uint16_t bits) {
                   ^
/Users/ehaas/source/zig/build/zig2.c:800411:30: error: call to undeclared function 'zig_lo_big'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   t4 = zig_wrap_u8((uint8_t)zig_lo_big(zig_shr_big(t9, UINT16_C(236))), UINT8_C(4));
                             ^
```

**Comment by ehaas** [comment]
@Vexu I will need your help integrating the arocc's `GenerateDef` changes with zig's `build.zig` - I'm not sure how to add the def files to the aro module.

**Comment by Vexu** [comment]
I don't think there's going to be an easy way to do so because of the bootstrapping process. I'd just copy the generated files from the cache and replace the def files.

**Comment by Vexu** [comment]
One possible option would be to add a stubbed def file and then in CMake add that as the module for the def files for the purpose of bootstrapping. For building them regularly it should be easy enough to copy some parts from Aro's `build.zig`.

A stub would look something like this:
```zig
pub fn with(comptime _: type) type {
    return struct {
        pub const max_param_count = 0;
        pub const longest_name = 0;
        pub const data = [_]@This(){.{}};
        pub fn fromName(_: []const u8) ?@This() {
            return null;
        }
        pub fn nameFromUniqueIndex(_: u16, _: []u8) []u8 {
            return "";
        }
        pub fn uniqueIndex(_: []const u8) ?u16 {
            return null;
        }
        pub const Tag = enum(u16) { _ };
        pub fn nameFromTag(_: Tag) NameBuf {}
        pub fn tagFromName(name: []const u8) ?Tag {
            return @enumFromInt(name.len);
        }
        pub const NameBuf = struct {
            pub fn span(_: *const NameBuf) []const u8 {
                return "";
            }
        };
    };
}

```

**Review by Vexu — APPROVED** [review]
I forgot about the "file exists in multiple modules" error again and made the stub work for both. It could be simplified for them individually but this works too.

**Comment by ehaas** [comment]
The issue I was seeing was that stage3 was unable to build the zig compiler due to not finding the def files. I fixed it with the following diff:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index dcbddc476..d5fd9b4b1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -647,8 +647,11 @@ set(ZIG_STAGE2_SOURCES
     "${CMAKE_SOURCE_DIR}/src/windows_sdk.zig"
     "${CMAKE_SOURCE_DIR}/src/stubs/aro_builtins.zig"
     "${CMAKE_SOURCE_DIR}/src/stubs/aro_attributes.zig"
+    "${CMAKE_SOURCE_DIR}/deps/aro/Attribute/names.def"
+    "${CMAKE_SOURCE_DIR}/deps/aro/Builtins/Builtin.def"
 )

+
 if(MSVC)
     set(MSVC_DIA_SDK_DIR "$ENV{VSINSTALLDIR}DIA SDK")
     if(IS_DIRECTORY ${MSVC_DIA_SDK_DIR})
@@ -947,3 +950,10 @@ install(CODE "set(ZIG_BUILD_ARGS \"${ZIG_BUILD_ARGS}\")")
 install(CODE "set(CMAKE_INSTALL_PREFIX \"${CMAKE_INSTALL_PREFIX}\")")
 install(CODE "set(CMAKE_SOURCE_DIR \"${CMAKE_SOURCE_DIR}\")")
 install(SCRIPT "${CMAKE_SOURCE_DIR}/cmake/install.cmake")
+
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/deps/aro/Attribute/names.def
+     DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/deps/aro/Attribute)
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/deps/aro/Builtins/Builtin.def
+     DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/deps/aro/Builtins)
```

I'm not sure if this is the correct fix; I also did this without your changes to this branch or to aro from today.

**Comment by andrewrk** [comment]
Please don't add any additional logic to CMakeLists.txt. Zig should depend on Aro only as a set of zig files. I don't want to add any Aro-related build system logic to zig's build process.

**Comment by Vexu** [comment]
> The issue I was seeing was that stage3 was unable to build the zig compiler due to not finding the def files.

I think the issue was https://github.com/Vexu/arocc/commit/9f4c28ae36b664238ce8e4ac1d0e7fd147d16726

**Comment by ehaas** [comment]
If we want to completely eliminate the `CMakeLists.txt` changes (so, no stub modules or anything) I have a local patch that does that by replacing the `.def` file contents with the generated versions. It succeeds at the bootstrap + stage4 build. Let me know if that is desired and I can push it.

**Comment by andrewrk** [comment]
Thanks @ehaas. I think this is fine - passing a few more args to building zig2.c is OK, I just didn't want to add any more steps to the process. For example, to update #17892 after merging this PR, I will only need to adjust a few command line args, but not introduce more logic into bootstrap.c.

**Pull request merged**